Add first blog post for advanced networking in AWS. (#2631)
pulumi/pulumi-hugo: themes/default/content/blog/advanced-aws-networking-part-1/index.md

# A placeholder image representing the recommended format, dimensions and aspect ratio
# has been provided for you.
meta_image: meta.png

authors:
    - josh-kodroff
    - andy-taylor
    - jose-juhala

tags:
    - aws
    - networking
    - hub-and-spoke
    - python
---

In this blog series we'll teach you how to create a hub-and-spoke network architecture in AWS with centralized egress. In this first installment, we'll talk about the benefits of this architecture and begin to lay out some of its components in Python with Pulumi, the infrastructure as code tool that enables you to manage infrastructure with real programming languages!

<!--more-->

{{% notes type="info" %}}
The complete code for this project is available on [GitHub](https://github.com/pulumi/examples/tree/master/aws-py-hub-and-spoke).
{{% /notes %}}

## What is a hub-and-spoke network?

A hub-and-spoke network is a common architecture for creating a network topology that provides isolation and security for your workloads. Our hub-and-spoke architecture on AWS has three main components: an inspection VPC, AWS Transit Gateway, and a series of spoke VPCs.

> * The inspection VPC provides centralized egress. It is the only VPC that can communicate with the internet, and all other VPCs must route their traffic through the inspection VPC. In a later blog post, we'll add AWS Firewall to the inspection VPC to allow all traffic to the internet to be governed by a single set of policies.

Author: It should be "inspection VPC" everywhere. I was actually using "hub" incorrectly IMO: the TGW is the hub.

> * The inspection VPC provides centralized egress. It is the only VPC that can communicate with the internet, and all other VPCs must route their traffic through the inspection VPC. In a later blog post, we'll add AWS Firewall to the inspection VPC to allow all traffic to the internet to be governed by a single set of policies.

Author: Also updated the top comment to link to the PR for the example code.

---
title: "Advanced AWS Networking, Part 1"

# The date represents the post's publish date,
# and by default corresponds with the date this file was generated.
# Posts with future dates are visible in development,
# but excluded from production builds.
# Use the time and timezone-offset portions of of this value
# to schedule posts for publishing later.

> date: 2023-03-28T11:17:23-04:00

Reviewer: ```suggestion
date: 2023-04-05
```

For simplicity, we will not be enabling routes between spoke VPCs in this architecture.
{{% /notes %}}

For a full discussion of all the considerations of building multi-VPC networks on AWS, see the excellent AWS whitepaper [Building a Scalable and Secure Multi-VPC AWS Network Infrastructure](https://docs.aws.amazon.com/whitepapers/latest/building-scalable-secure-multi-vpc-network-infrastructure/welcome.html).

## Why use a hub-and-spoke network with centralized egress?

There’s several advantages to this architecture:

* **Cost savings:** We save costs by allowing all of our spoke VPCs to share a single set of highly-available NAT gateways in the inspection VPC. There are costs to sending our traffic over AWS Transit Gateway, but in most scenarios Transit Gateway costs are significantly less than the cost of running NAT gateways in each spoke VPC.

> * **Stable public IP addresses:** Because all of our egress traffic to the internet originate from the same set of stable, public elastic IP addresses (attached to the NAT gateways in the inspection VPC), we can more easily interface with systems that use high-trust, network based security models (i.e., IP allowlists). Network administrators for the systems we want to interface with only need allow a limited number of IPs to establish connectivity, as opposed to having to allow new IPs for every VPC we create. These high-trust systems are often based in the data center and may have small and infrequent change windows, so only having to go through the change management process once, not for every spoke VPC we add, is a big win.

Author: Fixed (slightly different from suggested wording).

    "post-inspection-tgw-route-table",
    aws.ec2transitgateway.RouteTableArgs(
        transit_gateway_id=tgw.id,
        tags={
            "Name": "post-inspection",
        }
    ),
    # Adding the TGW as the parent makes the output of `pulumi up` a little
    # easier to understand as it groups these resources visually under the TGW
    # on which they depend.
    opts=pulumi.ResourceOptions(
        parent=tgw,
    ),
)

spoke_tgw_route_table = aws.ec2transitgateway.RouteTable(
    "spoke-tgw-route-table",
    aws.ec2transitgateway.RouteTableArgs(
        transit_gateway_id=tgw.id,
        tags={
            "Name": "spoke-tgw",
        }
    ),
    opts=pulumi.ResourceOptions(
        parent=tgw,
    ),
)
```

## Creating the Inspection VPC

{{% notes type="info" %}}
The code samples in this section are not exhaustive. That is, they do not contain every line of code required so that we can focus on the most illustrative portions. See the [GitHub repo]([GitHub](https://github.com/pulumi/examples/tree/master/aws-py-hub-and-spoke) for full context.
{{% /notes %}}

Now that you've added some foundational Transit Gateway resources, you can create your inspection VPC.

> The inspection VPC is implemented as a [Pulumi component resource](https://www.pulumi.com/docs/intro/concepts/resources/components/). While we do not need to create multiple inspection VPCs in this codebase, implementing the inspection VPC as a component resource enhances the readability of both the code and the output by grouping all resources that belong to the inspection VPC together in the output of `pulumi up`. You also can further improve the readability of the component by placing its code in a separate file (in this case, `inspection.py`).

Reviewer 2: ```suggestion
The inspection VPC is implemented as a [Pulumi component resource](https://www.pulumi.com/docs/intro/concepts/resources/components/). While we do not need to create multiple inspection VPCs in this codebase, implementing the inspection VPC as a component resource enhances the readability of both the code and the output by grouping all resources that belong to the inspection VPC in the output of `pulumi up`. You can also improve the component's readability by placing its code in a separate file (in this case, `inspection.py`).
```

    route_table = aws.ec2.get_route_table(
        subnet_id=subnet_id
    )

    aws.ec2.Route(
        f"{self.name}-route-{subnet_id}-to-tgw",
        aws.ec2.RouteArgs(
            route_table_id=route_table.id,
            destination_cidr_block="0.0.0.0/0",
            nat_gateway_id=self.nat_gateway.id,
        ),
        pulumi.ResourceOptions(
            parent=self,
        ),
    )
```

Finally, you'll need to add routes within the inspection VPC's public subnets (where the NAT gateways live) that route return traffic from the internet back through the Transit Gateway. This is how the inspection VPC will be able to route traffic to the spoke VPCs:

```python
for subnet_id in public_subnet_ids:
    route_table = aws.ec2.get_route_table(
        subnet_id=subnet_id
    )

    aws.ec2.Route(
        f"{self.name}-route-{subnet_id}-to-tgw",
        aws.ec2.RouteArgs(
            route_table_id=route_table.id,
            destination_cidr_block=self.args.supernet_cidr_block,
            transit_gateway_id=self.args.tgw_id,
        ),
        pulumi.ResourceOptions(
            depends_on=[self.tgw_attachment],
            parent=self,
        ),
    )
```

> You need to both sets of these routes within an [apply](https://www.pulumi.com/docs/intro/concepts/inputs-outputs/#apply) because of we need to access the raw values of each of the subnet ids. Apply can be conceptually difficult for beginners in Pulumi to grasp, but it essentially means "once you know all of the values that I need, pass them on to this function".

Reviewer 2: ```suggestion
You need both sets of these routes in an [apply](https://www.pulumi.com/docs/intro/concepts/inputs-outputs/#apply) because we need to access the raw values of each of the subnet ids. Apply means once you know all of the values needed, pass them on to this function.
```

Author: Fixed.

## Final Steps

Now that you've created our inspection VPC component, you can reference it in the main project file:

```python
from inspection import InspectionVpc, InspectionVpcArgs
```

and you can create an instance of the component:

```python
insp_vpc = InspectionVpc(
    "inspection",
    InspectionVpcArgs(
        supernet_cidr_block=supernet_cidr,
        vpc_cidr_block="10.129.0.0/24",
        tgw_id=tgw.id,
        inspection_tgw_route_table_id=inspection_tgw_route_table.id,
        spoke_tgw_route_table_id=spoke_tgw_route_table.id,
        firewall_policy_arn=None, # We will add this in the next post
    )
)
```

At this point, we do not have have a fully functional architecture - we still need to create the spoke VPCs - but if we want to get some feedback on our architecture, we can run `pulumi up` and see what happens. We should see that Pulumi will create the VPC, the Transit Gateway attachment, and the routes.

> In the following post, we’ll create our spoke VPCs and verify connectivity, then add centralized inspection of network traffic going to the internet. Stay tuned!

Reviewer 2: ```suggestion
In the next post, we’ll create our spoke VPCs and verify connectivity, then add centralized inspection of network traffic going to the internet. Stay tuned!
```
